20240129-satellite-managed-encryption.md

```diff
@@ -0,0 +1,160 @@
+* a credential to access the Secret Manager, in this case, a service account.
+  * the service account should have the minimum permissions required to access the Secret Manager.
+  * the service account should possibly be IP restricted to the satellite's IP address.
+  * the service account should be restricted to access only the specific secretes needed for this purpose; the Master Key.
+  * alternatively, a Kubernates Service account can be used in place of the service account so we do not actually expose the service account itself, while also restricting access to the Kubernetes pod.
+* the project ID of the project that the Secret Manager is in.
+* the secret ID of the master key, which at this point should already be created in the Secret Manager.
+
+Alternatively, we can use the [GCP Secret Store CSI driver](https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp) to mount the Master Key in the Satellite pod without having to expose any credentials at all.
+However, there are [security considerations](https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp?tab=readme-ov-file#security-considerations) we have to make.
+
+#### Satellite Project Updates
+
+A new column should be added to the `projects` table to be the passphrase store, like `projects.passphrase_enc`. This should be the encrypted passphrase for satellite-managed-encryption projects only and `null` for others.
+Projects that are created before this feature is deployed and projects created after but opt out should have `null` in this column.
+
+Functionality should then be added to allow creating a project with a satellite managed encryption passphrase:
+* the console service receives a request to create a project
+* the console service requests a master-key-encrypted encrypted, cryptographically random passphrase from the KMS service
+* the console service creates the project with the encrypted passphrase
+* endpoints on the satellite API are added or updated so that an authenticated user can request the encryption passphrase for a project during access creation
+  * endpoints handlers should use the KMS service to decrypt the project's encrypted
+
+#### Satellite UI Updates
+
+When creating a project, the user should be provided the option to create different types of projects:
+
+* satellite-managed-encryption project
+* user-managed-encryption project
+
+The user-managed-encryption project will be the same as before, and user will always be prompted to enter encryption passphrase when necessary, e.g. when using the file browser or generating an access grant.
+
+The satellite-managed-encryption project will never prompt the user for an encryption passphrase in the UI. Instead, when the user does an action that requires generating credentials, the UI will retrieve the project's encryption passphrase from the satellite KMS, and generate an access grant using this passphrase, and path encryption disabled.
```

Comment:
Yeah, technically they could be separate. The main reason they are related is because we decided that they were - they both are usability improvements, and they both involve updating the access encryption rules.
In the end, we will still want to present a binary decision to the user - either you select the user-managed encryption project that behaves like current projects (user provided passphrase + path encryption). Or you select the satellite-managed encryption project (satellite managed passphrase + no path encryption).
Behind the scenes, we can configure these two things separately, allowing us to toggle them independently in the future. But for the purposes of this design, they will be grouped together in the UI regardless of how it's configured behind the scenes.

Does that make sense? After thinking about your question, I do think it makes sense to have two separate DB flags, but I still think within the UI we should not worry about making that distinction for the initial work.

Reply:
If it's a product decision to package them together then that's good enough explanation for me.

Reply:
well I talked with Wilfred about it, and we actually think splitting the flags is a good idea, since it gives us more options in the future. So it was a good suggestion and we've integrated it into the doc - thanks!
We will still be grouping the two features together from a product perspective, but we do have the option now to, for example, have user-managed encryption but without path encryption.